### PR TITLE
Implement Chi2 distribution

### PR DESCRIPTION
**Pull request by alicanb**
@fritzo FYI for some reason df=0.01 case fails for `test_chi2_sample_grad`. I removed that case thinking it's probably a fluke (I tried different seeds, it didn't go away)
cc. @apaszke

### CODE REVIEW

**Comment by fritzo** [comment]
>  for some reason df=0.01 case fails for `test_chi2_sample_grad`

I think that's fine since very small `df<0.1` is seldom used in practice. Two possible causes are:
1.  the `standared_gamma()` sampler puts most mass near zero for `alpha<0.01`. This heavy zero might cause the tests to fail.
2. the `standard_gamma_grad()` function has worst-case relative error of around 1% which might cause test failure. @martinjankowiak has found a more precise implementation (based on Rice saddle point expansions) and we are planning to implement it in the next couple weeks.

File: docs/source/distributions.rst [hunk 1]
```diff
@@ -31,20 +31,44 @@ Probability distributions - torch.distributions
...
+.. autoclass:: Chi2
+    :members:
+
 :hidden:`Dirichlet`
 ~~~~~~~~~~~~~~~~~~~~~~~
 
 .. autoclass:: Dirichlet
     :members:
 
+:hidden:`Exponential`
```
Comment (fritzo) [inline docs/source/distributions.rst]: Thanks for updating documentation!

File: test/test_distributions.py [hunk 2]
```diff
@@ -480,6 +488,63 @@ def test_gamma_sample_grad(self):
...
+                                        scipy.stats.chi2(df),
+                                        'Chi2(df={})'.format(df))
+
+        # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_chi2_sample_grad(self):
+        self._set_rng_seed(1)
+        num_samples = 100
+        for df in [1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4]:
+            # NOTE: for some reason test fails for df = 1e-2
```
Comment (fritzo) [inline test/test_distributions.py]: I've created an issue for this and @martinjankowiak and I will see if we can fix it in the next couple weeks https://github.com/probtorch/pytorch/issues/48

**Review by fritzo — APPROVED** [review]
LGTM

File: torch/distributions/__init__.py [hunk 3]
```diff
@@ -39,6 +39,8 @@
 from .normal import Normal
 from .exponential import Exponential
 from .laplace import Laplace
+from .chi2 import Chi2
```
Comment (fritzo) [inline torch/distributions/__init__.py]: @alicanb Would you mind importing `Uniform` in this PR to help clean up?
```py
from .uniform import Uniform
```
since it was missed in #4328 ?

File: test/test_distributions.py [hunk 4]
```diff
@@ -7,7 +7,7 @@
 import torch
 from common import TestCase, run_tests
 from torch.autograd import Variable, gradcheck
-from torch.distributions import (Bernoulli, Beta, Categorical, Dirichlet,
+from torch.distributions import (Bernoulli, Beta, Categorical, Chi2, Dirichlet,
                                  Exponential, Gamma, Laplace, Normal)
 from torch.distributions.uniform import Uniform
```
Comment (fritzo) [inline test/test_distributions.py]: Would you mind cleaning up the `Uniform` import, moving it into `from torch.distributions import ...`?

**Review by fritzo — COMMENTED** [review]
Would you mind cleaning up some unrelated `Uniform` import issues in this PR (to avoid an extra nuisance PR)? If that's a pain, we can fix them as part of `Cauchy` or something.

**Comment by alicanb** [comment]
Sure I'm on my way to airport, I think I'll have plenty of time to fix that

**Comment by alicanb** [comment]
@apaszke mind taking a look?

File: test/test_distributions.py [hunk 5]
```diff
@@ -522,6 +530,63 @@ def test_gamma_sample_grad(self):
...
+            d = df.data.view(-1)[idx]
+            expected = scipy.stats.chi2.logpdf(x, d)
+            self.assertAlmostEqual(log_prob, expected, places=3)
+
+        self._check_log_prob(Chi2(df), ref_log_prob)
+
+    # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_chi2_sample(self):
+        set_rng_seed(0)
```
Comment (apaszke) [inline test/test_distributions.py]: This needs a reference to the note about testing that has been added in one of the recent PRs. @fritzo Does it also apply to the test below?

File: test/test_distributions.py [hunk 6]
```diff
@@ -522,6 +530,63 @@ def test_gamma_sample_grad(self):
...
+                                        scipy.stats.chi2(df),
+                                        'Chi2(df={})'.format(df))
+
+        # This is a randomized test.
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_chi2_sample_grad(self):
+        set_rng_seed(1)
+        num_samples = 100
+        for df in [1e-3, 1e-1, 1e0, 1e1, 1e2, 1e3, 1e4]:
+            # NOTE: for some reason test fails for df = 1e-2
```
Comment (apaszke) [inline test/test_distributions.py]: I'm a bit worried about this failure. Any ideas what might be causing it? Numerical instability? Unlucky seed?

Comment (fritzo) [reply test/test_distributions.py]: I suspect this is due to underflow in the underlying `Gamma` sampler: when `alpha<=0.01` lots of mass ends up at zero (or near zero #4262). I'm even surprised the 1e-3 case passes. For example [these plots](https://github.com/fritzo/notebooks/blob/master/debug-gamma-sampler.ipynb) show the difference between the double-precision scipy sampler and single-precision `torch.distributions.Gamma` sampler. Maybe this test will pass if you use `DoubleTensor` instead of `FloatTensor`?

I doubt this numerical instability it's an issue in practice since `Chi2` is usually created with `df >= 1` (at least in my experience).

File: test/test_distributions.py (hunk 5 above)
Comment (fritzo) [reply test/test_distributions.py]: Yes. @alicanb In #4357 I've updated all comments as suggested by Edward:
```diff
- # This is a randomized test.
  @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
  def test_chi2_sample(self):
-    set_rng_seed(0)
+    set_rng_seed(0)  # see Note [Randomized statistical tests]
```

File: test/test_distributions.py (hunk 6 above)
Comment (fritzo) [reply test/test_distributions.py]: We're tracking this issue at probtorch#48

Comment (apaszke) [reply test/test_distributions.py]: What worries me is not so much the underflow, as the fact that it passes with 1e-3. But considering your notebook and the fact that it's actually a test that compares to scipy, it might be the case that at values that low they both underflow horribly and compare almost equal 😄 

How about putting up a warning just in case?

Comment (alicanb) [reply test/test_distributions.py]: I just tried df=0.01, it works now. However there's something wrong with the branch I can't build it locally (I have no idea how it passes CI test). I will create a new PR.
